File: android/app/src/main/java/com/studyapp/domain/usecase/GetHomeDataUseCase.kt
package com.studyapp.domain.usecase

import com.studyapp.domain.model.Exam
import com.studyapp.domain.model.Goal
import com.studyapp.domain.model.GoalType
import com.studyapp.domain.model.StudySession
import com.studyapp.domain.model.StudyWeekday
import com.studyapp.domain.model.TimetableEntry
import com.studyapp.domain.model.TimetableLesson
import com.studyapp.domain.model.TimetablePeriod
import com.studyapp.domain.model.TimetableTerm
import com.studyapp.domain.model.TodayReviewProblem
import com.studyapp.domain.model.latestActiveDailyGoal
import com.studyapp.domain.model.latestActiveWeeklyGoal
import com.studyapp.domain.repository.ExamRepository
import com.studyapp.domain.repository.GoalRepository
import com.studyapp.domain.repository.MaterialRepository
import com.studyapp.domain.repository.ProblemReviewRepository
import com.studyapp.domain.repository.StudySessionRepository
import com.studyapp.domain.repository.SubjectRepository
import com.studyapp.domain.repository.TimetableRepository
import com.studyapp.domain.util.Clock
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import java.time.LocalDate
import java.time.LocalTime
import javax.inject.Inject

data class HomeData(
    val todayStudyMinutes: Long,
    val todaySessions: List<TodaySession>,
    val todayGoal: Goal?,
    val weeklyGoal: Goal?,
    val weeklyStudyMinutes: Long,
    val upcomingExams: List<Exam>,
    val timetableLesson: TimetableLesson? = null,
    val todayReviewProblems: List<TodayReviewProblem> = emptyList()
)

data class TodaySession(
    val id: Long,
    val subjectName: String,
    val materialName: String,
    val duration: Long,
    val startTime: Long
)

class GetHomeDataUseCase @Inject constructor(
    private val studySessionRepository: StudySessionRepository,
    private val materialRepository: MaterialRepository,
    private val subjectRepository: SubjectRepository,
    private val problemReviewRepository: ProblemReviewRepository,
    private val goalRepository: GoalRepository,
    private val examRepository: ExamRepository,
    private val timetableRepository: TimetableRepository,
    private val clock: Clock
) {
    operator fun invoke(): Flow<HomeData> {
        val todayStart = clock.startOfToday()
        val todayEnd = todayStart + DAY_MS - 1
        val weekStart = clock.startOfWeek()
        val todayStudyWeekday = StudyWeekday.fromDayOfWeek(clock.currentLocalDate().dayOfWeek)

        val todaySessionsFlow: Flow<List<StudySession>> = studySessionRepository
            .getSessionsBetweenDates(todayStart, todayStart + DAY_MS)
            .map { result -> result.getOrNull() ?: emptyList() }

        val todayStudyMinutesFlow: Flow<Long> = todaySessionsFlow.map { sessions ->
            sessions.sumOf { it.duration / 60000 }
        }

        val todaySessionsMappedFlow: Flow<List<TodaySession>> = todaySessionsFlow.map { sessions ->
            sessions.map { session ->
                TodaySession(
                    id = session.id,
                    subjectName = session.subjectName,
                    materialName = session.materialName,
                    duration = session.duration,
                    startTime = session.startTime
                )
            }
        }

        val allGoalsFlow = goalRepository.getAllGoals()
            .map { result -> result.getOrNull() ?: emptyList() }

        val todayGoalFlow: Flow<Goal?> = allGoalsFlow.map { goals ->
            goals.latestActiveDailyGoal(todayStudyWeekday)
        }

        val weeklyGoalFlow: Flow<Goal?> = allGoalsFlow.map { goals ->
            goals.latestActiveWeeklyGoal()
        }

        val weeklyStudyMinutesFlow: Flow<Long> = studySessionRepository
            .getSessionsBetweenDates(weekStart, weekStart + WEEK_MS)
            .map { result -> result.getOrNull() ?: emptyList() }
            .map { sessions ->
                sessions.sumOf { it.duration / 60000 }
            }

        val upcomingExamsFlow: Flow<List<Exam>> = examRepository.getUpcomingExams()
            .map { result -> result.getOrNull() ?: emptyList() }

        val todayReviewProblemsFlow: Flow<List<TodayReviewProblem>> = combine(
            problemReviewRepository.getActiveReviewRecords().map { it.getOrNull() ?: emptyList() },
            materialRepository.getAllMaterials().map { it.getOrNull() ?: emptyList() },
            subjectRepository.getAllSubjects().map { it.getOrNull() ?: emptyList() }
        ) { reviews, materials, subjects ->
            val materialMap = materials
                .filter { it.deletedAt == null }
                .associateBy { it.id }
            val subjectMap = subjects
                .filter { it.deletedAt == null }
                .associateBy { it.id }
            reviews
                .groupBy { it.problemId.ifBlank { com.studyapp.domain.model.ProblemReviewRecord.problemId(it.materialId, it.problemNumber) } }
                .mapNotNull { (_, problemReviews) ->
                    problemReviews.maxByOrNull { it.reviewedAt }
                }
                .filter { it.nextReviewDate <= todayEnd && it.deletedAt == null }
                .mapNotNull { review ->
                    val material = materialMap[review.materialId] ?: return@mapNotNull null
                    val subject = subjectMap[material.subjectId]
                    TodayReviewProblem(
                        materialId = material.id,
                        materialName = material.name,
                        subjectName = subject?.name.orEmpty(),
                        problemNumber = review.problemNumber,
                        nextReviewDate = review.nextReviewDate,
                        consecutiveCorrectCount = review.consecutiveCorrectCount,
                        wrongCount = review.wrongCount
                    )
                }
                .sortedWith(
                    compareBy<TodayReviewProblem> { it.nextReviewDate }
                        .thenBy { it.materialName }
                        .thenBy { it.problemNumber }
                )
        }

        val timetableLessonFlow: Flow<TimetableLesson?> = combine(
            timetableRepository.getAllPeriods().map { it.getOrNull() ?: emptyList() },
            timetableRepository.getAllEntries().map { it.getOrNull() ?: emptyList() },
            timetableRepository.getAllTerms().map { it.getOrNull() ?: emptyList() }
        ) { periods, entries, terms ->
            nextTimetableLesson(periods, entries, terms, LocalDate.now())
        }

        val todayDataFlow: Flow<Triple<Long, List<TodaySession>, Goal?>> = combine(
            todayStudyMinutesFlow,
            todaySessionsMappedFlow,
            todayGoalFlow
        ) { todayMinutes, todaySessions, todayGoal ->
            Triple(todayMinutes, todaySessions, todayGoal)
        }

        data class SecondaryHomeData(
            val weeklyGoal: Goal?,
            val weeklyMinutes: Long,
            val exams: List<Exam>,
            val timetableLesson: TimetableLesson?,
            val todayReviewProblems: List<TodayReviewProblem>
        )

        val secondaryHomeDataFlow: Flow<SecondaryHomeData> = combine(
            weeklyGoalFlow,
            weeklyStudyMinutesFlow,
            upcomingExamsFlow,
            timetableLessonFlow,
            todayReviewProblemsFlow
        ) { weeklyGoal, weeklyMinutes, exams, timetableLesson, todayReviewProblems ->
            SecondaryHomeData(
                weeklyGoal = weeklyGoal,
                weeklyMinutes = weeklyMinutes,
                exams = exams,
                timetableLesson = timetableLesson,
                todayReviewProblems = todayReviewProblems
            )
        }

        return combine(
            todayDataFlow,
            secondaryHomeDataFlow
        ) { todayData, secondaryData ->
            val (todayMinutes, todaySessions, todayGoal) = todayData
            HomeData(
                todayStudyMinutes = todayMinutes,
                todaySessions = todaySessions.sortedByDescending { it.startTime },
                todayGoal = todayGoal,
                weeklyGoal = secondaryData.weeklyGoal,
                weeklyStudyMinutes = secondaryData.weeklyMinutes,
                upcomingExams = secondaryData.exams.sortedBy { it.date },
                timetableLesson = secondaryData.timetableLesson,
                todayReviewProblems = secondaryData.todayReviewProblems
            )
        }
    }

    private fun nextTimetableLesson(
        periods: List<TimetablePeriod>,
        entries: List<TimetableEntry>,
        terms: List<TimetableTerm>,
        reference: LocalDate
    ): TimetableLesson? {
        val activePeriods = periods
            .filter { it.isActive && it.deletedAt == null && it.startMinute < it.endMinute }
            .sortedWith(compareBy<TimetablePeriod> { it.sortOrder }.thenBy { it.startMinute })
        if (activePeriods.isEmpty()) return null

        val periodMap = activePeriods.associateBy { it.id }
        val activeTerm = terms.firstOrNull { it.deletedAt == null && it.isActive && it.contains(reference) }
            ?: terms.filter { it.deletedAt == null && it.isActive }.maxByOrNull { it.endDate }
        val referenceDay = reference.toEpochDay()
        val activeEntries = entries.filter {
            it.deletedAt == null &&
            (it.termId == activeTerm?.id || it.termId == null) &&
            (it.validFromDate?.let { d -> referenceDay >= d } ?: true) &&
            (it.validToDate?.let { d -> referenceDay <= d } ?: true) &&
            StudyWeekday.timetableDays.contains(it.dayOfWeek) &&
            periodMap[it.periodId] != null
        }
        if (activeEntries.isEmpty()) return null

        val now = LocalTime.now()
        val currentMinutes = now.hour * 60 + now.minute
        val referenceWeekday = StudyWeekday.fromDayOfWeek(reference.dayOfWeek)

        for (offset in 0 until 7) {
            val date = reference.plusDays(offset.toLong())
            val day = StudyWeekday.fromDayOfWeek(date.dayOfWeek)
            if (!StudyWeekday.timetableDays.contains(day)) continue

            val dayEntries = activeEntries
                .filter { it.dayOfWeek == day }
                .mapNotNull { entry ->
                    val period = periodMap[entry.periodId] ?: return@mapNotNull null
                    Pair(entry, period)
                }
                .sortedBy { it.second.startMinute }

            for (pair in dayEntries) {
                val entry = pair.first
                val period = pair.second
                if (offset == 0 && day == referenceWeekday) {
                    if (currentMinutes >= period.startMinute && currentMinutes < period.endMinute) {
                        return TimetableLesson(entry = entry, period = period, dayOfWeek = day, date = date, isCurrent = true)
                    }
                    if (currentMinutes >= period.endMinute) continue
                }
                return TimetableLesson(entry = entry, period = period, dayOfWeek = day, date = date, isCurrent = false)
            }
        }

        return null
    }

    companion object {
        private const val DAY_MS = 24 * 60 * 60 * 1000L
        private const val WEEK_MS = 7 * DAY_MS
    }
}

File: ios/StudyApp/Domain/UseCases/GetHomeDataUseCase.swift
import Foundation

struct GetHomeDataUseCase {
    let studySessionRepository: StudySessionRepository
    let goalRepository: GoalRepository
    let examRepository: ExamRepository
    let timetableRepository: TimetableRepository
    let problemReviewRepository: ProblemReviewRepository
    let clock: Clock

    func execute() async throws -> HomeData {
        let todayStart = clock.startOfToday()
        let weekStart = clock.startOfWeek()
        let todayWeekday = StudyWeekday.from(calendarWeekday: Calendar.current.component(.weekday, from: clock.now()))
        let dayMs: Int64 = 86_400_000
        let weekMs = dayMs * 7

        async let todaySessionsTask = studySessionRepository.getSessionsBetweenDates(start: todayStart, end: todayStart + dayMs)
        async let goalsTask = goalRepository.getAllGoals()
        async let weeklySessionsTask = studySessionRepository.getSessionsBetweenDates(start: weekStart, end: weekStart + weekMs)
        async let upcomingExamsTask = examRepository.getUpcomingExams(now: clock.now())
        async let timetablePeriodsTask = timetableRepository.getAllTimetablePeriods()
        async let timetableEntriesTask = timetableRepository.getAllTimetableEntries()
        async let timetableTermsTask = timetableRepository.getAllTimetableTerms()
        async let todayReviewProblemsTask = problemReviewRepository.getTodayReviewProblems(reference: clock.now())

        let todaySessions = try await todaySessionsTask
        let goals = try await goalsTask
        let weeklySessions = try await weeklySessionsTask
        let upcomingExams = try await upcomingExamsTask
        let timetablePeriods = try await timetablePeriodsTask
        let timetableEntries = try await timetableEntriesTask
        let timetableTerms = try await timetableTermsTask
        let todayReviewProblems = try await todayReviewProblemsTask
        let todayGoal = goals.latestActiveDailyGoal(for: todayWeekday)
        let weeklyGoal = goals.latestActiveWeeklyGoal()

        let timetableLessons = nextTimetableLessons(
            periods: timetablePeriods,
            entries: timetableEntries,
            terms: timetableTerms,
            reference: clock.now()
        )
        let currentLesson = timetableLessons.first { $0.isCurrent }
        let upcomingLesson = timetableLessons.first { !$0.isCurrent }

        return HomeData(
            todayStudyMinutes: todaySessions.reduce(0) { $0 + $1.durationMinutes },
            todaySessions: todaySessions
                .sorted { $0.startTime > $1.startTime }
                .map {
                    TodaySession(
                        id: $0.id,
                        subjectName: $0.subjectName,
                        materialName: $0.materialName,
                        duration: $0.duration,
                        startTime: $0.startTime
                    )
            },
            todayGoal: todayGoal,
            weeklyGoal: weeklyGoal,
            weeklyStudyMinutes: weeklySessions.reduce(0) { $0 + $1.durationMinutes },
            upcomingExams: upcomingExams.sorted { $0.date < $1.date },
            timetableLesson: currentLesson,
            upcomingTimetableLesson: upcomingLesson,
            todayReviewProblems: todayReviewProblems
        )
    }

    private func nextTimetableLessons(
        periods: [TimetablePeriod],
        entries: [TimetableEntry],
        terms: [TimetableTerm],
        reference: Date
    ) -> [TimetableLesson] {
        let activePeriods = periods
            .filter { $0.isActive && $0.deletedAt == nil && $0.startMinute < $0.endMinute }
            .sorted { $0.sortOrder == $1.sortOrder ? $0.startMinute < $1.startMinute : $0.sortOrder < $1.sortOrder }
        guard !activePeriods.isEmpty else { return [] }

        let periodMap = Dictionary(uniqueKeysWithValues: activePeriods.map { ($0.id, $0) })
        let activeTerm = terms.first(where: { $0.deletedAt == nil && $0.isActive && $0.contains(reference) })
            ?? terms.filter { $0.deletedAt == nil && $0.isActive }.sorted { $0.endDate > $1.endDate }.first
        let referenceDay = reference.startOfDay.epochDay
        let activeEntries = entries.filter {
            $0.deletedAt == nil &&
            ($0.termId == activeTerm?.id || $0.termId == nil) &&
            ($0.validFromDate.map { referenceDay >= $0 } ?? true) &&
            ($0.validToDate.map { referenceDay <= $0 } ?? true) &&
            StudyWeekday.timetableDays.contains($0.dayOfWeek) &&
            periodMap[$0.periodId] != nil
        }
        guard !activeEntries.isEmpty else { return [] }

        let calendar = Calendar.current
        let currentMinutes = (calendar.component(.hour, from: reference) * 60) + calendar.component(.minute, from: reference)
        let referenceWeekday = StudyWeekday.from(calendarWeekday: calendar.component(.weekday, from: reference))
        var lessons: [TimetableLesson] = []

        for offset in 0..<7 {
            guard let date = calendar.date(byAdding: .day, value: offset, to: reference) else { continue }
            let day = StudyWeekday.from(calendarWeekday: calendar.component(.weekday, from: date))
            guard StudyWeekday.timetableDays.contains(day) else { continue }

            let dayEntries = activeEntries
                .filter { $0.dayOfWeek == day }
                .compactMap { entry -> (TimetableEntry, TimetablePeriod)? in
                    guard let period = periodMap[entry.periodId] else { return nil }
                    return (entry, period)
                }
                .sorted { $0.1.startMinute < $1.1.startMinute }

            for pair in dayEntries {
                let entry = pair.0
                let period = pair.1
                if offset == 0, day == referenceWeekday {
                    if currentMinutes >= period.startMinute && currentMinutes < period.endMinute {
                        lessons.append(TimetableLesson(entry: entry, period: period, dayOfWeek: day, date: date, isCurrent: true))
                        continue
                    }
                    if currentMinutes >= period.endMinute {
                        continue
                    }
                }
                lessons.append(TimetableLesson(entry: entry, period: period, dayOfWeek: day, date: date, isCurrent: false))
                if lessons.contains(where: { $0.isCurrent }) && lessons.contains(where: { !$0.isCurrent }) {
                    return lessons
                }
                if !lessons.contains(where: { $0.isCurrent }) && lessons.count >= 1 {
                    return lessons
                }
            }
        }

        return lessons
    }
}
